File: py/telemetry/odometer.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from telemetry import odometer_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class Tyre:
    id: str = ""
    kilometers: float = 0.0
    
    _proto_message: odometer_pb2.Tyre = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = odometer_pb2.Tyre()

    def _populate_proto(self):
        self._proto_message.id = self.id
        self._proto_message.kilometers = self.kilometers

    @classmethod
    def _from_proto(cls, proto_message) -> "Tyre":
        return cls(
            id = proto_message.id,
            kilometers = proto_message.kilometers,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Tyre":
        message = odometer_pb2.Tyre()
        message.ParseFromString(data)
        return cls(
            id = message.id,
            kilometers = message.kilometers,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Tyre":
        message = odometer_pb2.Tyre()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class Odometer:
    startDate: str = ""
    fl: Tyre = None
    fr: Tyre = None
    rl: Tyre = None
    rr: Tyre = None
    
    _proto_message: odometer_pb2.Odometer = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = odometer_pb2.Odometer()

    def _populate_proto(self):
        self._proto_message.startDate = self.startDate
        if self.fl:
            self.fl._populate_proto()
            self._proto_message.fl.CopyFrom(self.fl._proto_message)
        if self.fr:
            self.fr._populate_proto()
            self._proto_message.fr.CopyFrom(self.fr._proto_message)
        if self.rl:
            self.rl._populate_proto()
            self._proto_message.rl.CopyFrom(self.rl._proto_message)
        if self.rr:
            self.rr._populate_proto()
            self._proto_message.rr.CopyFrom(self.rr._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "Odometer":
        return cls(
            startDate = proto_message.startDate,
            fl = Tyre._from_proto(proto_message.fl),
            fr = Tyre._from_proto(proto_message.fr),
            rl = Tyre._from_proto(proto_message.rl),
            rr = Tyre._from_proto(proto_message.rr),
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Odometer":
        message = odometer_pb2.Odometer()
        message.ParseFromString(data)
        return cls(
            startDate = message.startDate,
            fl = (
                Tyre._from_proto(message.fl)
                if message.HasField("fl")
                else None
            ),
            fr = (
                Tyre._from_proto(message.fr)
                if message.HasField("fr")
                else None
            ),
            rl = (
                Tyre._from_proto(message.rl)
                if message.HasField("rl")
                else None
            ),
            rr = (
                Tyre._from_proto(message.rr)
                if message.HasField("rr")
                else None
            ),
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Odometer":
        message = odometer_pb2.Odometer()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/telemetry/odometer.cpp
#include "telemetry/odometer.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Telemetry
{
Tyre::Tyre(const PbTelemetry::Tyre& protobuf) {
    id = protobuf.id();
    kilometers = protobuf.kilometers();
}

Tyre::operator PbTelemetry::Tyre() const {
    PbTelemetry::Tyre ret;
    ret.set_id(id);
    ret.set_kilometers(kilometers);
    return ret;
}

std::string Tyre::serializeAsJsonString() const {
    PbTelemetry::Tyre protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Tyre::serializeAsProtobufString() const {
    PbTelemetry::Tyre protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Tyre::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::Tyre protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Tyre::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::Tyre protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

Odometer::Odometer(const PbTelemetry::Odometer& protobuf) {
    startDate = protobuf.startdate();
    fl = protobuf.fl();
    fr = protobuf.fr();
    rl = protobuf.rl();
    rr = protobuf.rr();
}

Odometer::operator PbTelemetry::Odometer() const {
    PbTelemetry::Odometer ret;
    ret.set_startdate(startDate);
    *(ret.mutable_fl()) = fl;
    *(ret.mutable_fr()) = fr;
    *(ret.mutable_rl()) = rl;
    *(ret.mutable_rr()) = rr;
    return ret;
}

std::string Odometer::serializeAsJsonString() const {
    PbTelemetry::Odometer protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Odometer::serializeAsProtobufString() const {
    PbTelemetry::Odometer protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Odometer::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::Odometer protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Odometer::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::Odometer protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}